cocotb/bfmgen.py

```diff
@@ -0,0 +1,398 @@
+    template_p = {
+        "bfm_name" : bfm_name,
+        "bfm_classname" : info.T.__module__ + "." + info.T.__qualname__,
+        }
+    
+    template = '''
+int ${bfm_name}_process_msg() __attribute__((weak));
+
+// Stub definition to handle the case where a referenced
+// BFM type isn't instanced
+int ${bfm_name}_process_msg() { }
+
+static void ${bfm_name}_notify_cb(void *user_data) {
+    svSetScope(user_data);
+    ${bfm_name}_process_msg();
+}
+
+int ${bfm_name}_register(const char *inst_name) {
+    return cocotb_bfm_register(
+        \"XXXX\", 
+        inst_name, 
+        \"${bfm_classname}\", 
+        &${bfm_name}_notify_cb, 
+        svGetScope());
+}
+'''
+    
+    return Template(template).safe_substitute(template_p)
+       
+def bfm_generate_sv(args):
+    inst = BfmMgr.inst()
+   
+    filename_c = args.o
+    if filename_c.find('.') != -1:
+        filename_c = os.path.splitext(filename_c)[0]
+    filename_c += ".c"
+    
+    out_sv = open(args.o, "w")
+    out_sv.write("//***************************************************************************\n")
+    out_sv.write("//* BFMs file for CocoTB. \n")
+    out_sv.write("//* Note: This file is generated. Do Not Edit\n")
+    out_sv.write("//***************************************************************************\n")
+
+    out_c = open(filename_c, "w")
+    out_c.write("//***************************************************************************\n")
+    out_c.write("//* BFMs DPI interface file for CocoTB. \n")
+    out_c.write("//* Note: This file is generated. Do Not Edit\n")
+    out_c.write("//***************************************************************************\n")
+    out_c.write("#include <stdio.h>\n")
+    out_c.write("#ifdef __cplusplus\n")
+    out_c.write("extern \"C\" {\n")
+    out_c.write("#endif\n")
+    out_c.write("\n")
+    out_c.write("#include \"svdpi.h\"\n")
+    out_c.write("typedef void (*cocotb_bfm_notify_f)(void *);\n")
+    out_c.write("int cocotb_bfm_register(const char *, const char *, const char *, cocotb_bfm_notify_f, void *);\n")
+
+    for t in inst.bfm_type_info_m.keys():
+        info = inst.bfm_type_info_m[t]
+        
+        if cocotb.bfm_vlog not in info.hdl.keys():
+            raise Exception("BFM \"" + t.__name__ + "\" does not support Verilog")
+        
+        template_f = open(info.hdl[cocotb.bfm_sv], "r")
+        template = template_f.read()
+        template_f.close()
+        
+        bfm_name = os.path.basename(info.hdl[cocotb.bfm_sv])
+    
+        if bfm_name.find('.') != -1:
+            bfm_name = os.path.splitext(bfm_name)[0]
+        
+        out_sv.write(process_template_sv(template, bfm_name, info))
+        out_c.write(generate_dpi_c(bfm_name, info))
+        
+    out_c.write("#ifdef __cplusplus\n")
+    out_c.write("}\n")
+    out_c.write("#endif\n")
+    
+    out_sv.close() 
+    out_c.close()
+
+def bfm_generate(args):
+    '''
+    Generates BFM files required for simulation
+    '''
```

Comment:
Nit: Write this as """Generate BFM files required for simulation.""" (https://www.python.org/dev/peps/pep-0257/#id16)